src/frequenz/sdk/data_ingestion/resampling/timeseries_resampling_actor.py

```diff
@@ -0,0 +1,159 @@
+                of a (component_id, metric_id) pair and `start_time`, which is the
+                point in time from which to subscribe  for data or `None` to subscribe
+                for data from the live component data stream
+
+        Returns:
+            receiver for receiving samples for a specific time series id
+        """
+        if subscription.time_series_id not in self._output_channels:
+            response: SubscriptionResponse = self._data_provider.subscribe(
+                time_series_id=subscription.time_series_id,
+                # Ignore parquet for now [later => start_time=subscription.start_time]
+                start_time=None,
+            )
+            self._input_receivers[subscription.time_series_id] = response.receiver
+            self._output_channels[subscription.time_series_id] = Broadcast[
+                Optional[Sample]
+            ](str(subscription.time_series_id))
+
+        return self._output_channels[subscription.time_series_id].get_receiver()
+
+    async def run(self) -> None:
+        """Run the actor."""
+        while True:
+            component_data_receivers = {
+                str(time_series_id): receiver
+                for time_series_id, receiver in self._input_receivers.items()
+            }
+
+            select = Select(
+                resampling_timer=self._resampling_timer,
+                subscription_receiver=self._subscription_receiver,
+                component_data_receiver=MergeNamed(**component_data_receivers),
+            )
+            while await select.ready():
+                if msg := select.subscription_receiver:
+                    subscription_request: SubscriptionRequest = msg.inner
+                    sample_receiver: Receiver[Optional[Sample]] = self.subscribe(
+                        subscription=subscription_request.subscription
+                    )
+                    await subscription_request.sender.send(
+                        SubscriptionResponse(
+                            receiver=sample_receiver,
+                            subscription=subscription_request.subscription,
+                        )
+                    )
+                    # Breaking out from the loop is required to regenerate
+                    # component_data_receivers to be able to fulfil this
+                    # subscription (later can be optimized by checking if
+                    # an output channel already existed in the `subscribe()` method)
+                    break
+                if _ := select.resampling_timer:
+                    samples: Dict[
+                        TimeSeriesId, Optional[Sample]
+                    ] = self._resampler.resample()
+                    for time_series_id, sample in samples.items():
+                        await self._output_channels[time_series_id].get_sender().send(
+                            sample
+                        )
```

Comment:
If you look at Broadcast channel implementation you will see that it creates new Sender with every get_sender call.
Please change your code to get_sender only once.

Reply:
Also, we probably shouldn't `await` on every `send()`, as we want to parallelize the sending. Although with the current implementation I guess if we don't await we'll just create all the new samples and just write it to the sender buffer but no receiver will be able to process anything until we await, right @sahas-subramanian-frequenz .

But when things get really concurrent, we should probably do something like:

```py
await async.gather(self._output_senders[time_series_id].send(sample)
    for time_series_id, sample in self._resampler.resample())
```

Reply:
Done